File: fixtures/node-application-owned/test/application.test.mjs
import assert from "node:assert/strict";
import http from "node:http";
import net from "node:net";
import test from "node:test";

import { createApplication } from "../src/application.mjs";

function getText(url) {
  return new Promise((resolve, reject) => {
    const request = http.get(url, { agent: false }, (response) => {
      let body = "";
      response.setEncoding("utf8");
      response.on("data", (chunk) => {
        body += chunk;
      });
      response.on("end", () => resolve({ status: response.statusCode, body }));
    });
    request.once("error", reject);
  });
}

async function waitFor(predicate, message) {
  const deadline = Date.now() + 2_000;
  while (!predicate()) {
    assert.ok(Date.now() < deadline, message);
    await new Promise((resolve) => setTimeout(resolve, 10));
  }
}

function connectionFails(port) {
  return new Promise((resolve) => {
    const socket = net.createConnection({ host: "127.0.0.1", port });
    socket.once("connect", () => {
      socket.destroy();
      resolve(false);
    });
    socket.once("error", () => resolve(true));
    socket.setTimeout(500, () => {
      socket.destroy();
      resolve(true);
    });
  });
}

test("shutdown is idempotent, drains an active request, and releases resources", async () => {
  const records = [];
  const application = createApplication({
    logger: (record) => records.push(record),
    applicationShutdownMs: 1_000,
  });
  const address = await application.start();
  const origin = `http://127.0.0.1:${address.port}`;

  const health = await getText(`${origin}/health`);
  assert.equal(health.status, 200);

  const activeRequest = getText(`${origin}/hold?ms=150`);
  await waitFor(
    () => application.metrics().activeRequests === 1,
    "the hold request never became active",
  );

  const firstShutdown = application.shutdown("TEST");
  const secondShutdown = application.shutdown("TEST");
  assert.strictEqual(firstShutdown, secondShutdown);

  await assert.rejects(getText(`${origin}/health`));
  const held = await activeRequest;
  assert.deepEqual(held, { status: 200, body: "held" });
  await firstShutdown;

  assert.deepEqual(application.metrics(), {
    ready: false,
    activeRequests: 0,
    cleanupRuns: 1,
    keepAliveActive: false,
    listening: false,
  });
  assert.strictEqual(application.shutdown("TEST"), firstShutdown);
  assert.equal(await connectionFails(address.port), true);
  assert.deepEqual(
    records.map((record) => record.event),
    [
      "server_ready",
      "shutdown_started",
      "resource_cleanup_completed",
      "shutdown_completed",
    ],
  );
});

File: fixtures/go-application-owned/internal/application/application_test.go
package application

import (
	"io"
	"net"
	"net/http"
	"reflect"
	"sync"
	"testing"
	"time"
)

func TestShutdownIsIdempotentDrainsRequestAndReleasesResources(t *testing.T) {
	application := New("127.0.0.1", 0, nil)
	address, err := application.Start()
	if err != nil {
		t.Fatal(err)
	}
	origin := "http://" + address.String()
	client := &http.Client{Transport: &http.Transport{DisableKeepAlives: true}}
	response, err := client.Get(origin + "/health")
	if err != nil || response.StatusCode != http.StatusOK {
		t.Fatalf("health request failed: status=%v err=%v", responseStatus(response), err)
	}
	_ = response.Body.Close()

	holdDone := make(chan string, 1)
	go func() {
		response, requestErr := client.Get(origin + "/hold?ms=150")
		if requestErr != nil {
			holdDone <- "error: " + requestErr.Error()
			return
		}
		body, _ := io.ReadAll(response.Body)
		_ = response.Body.Close()
		holdDone <- string(body)
	}()
	waitUntil(t, func() bool { return application.Metrics().ActiveRequests == 1 })

	results := make(chan error, 2)
	var callers sync.WaitGroup
	callers.Add(2)
	for caller := 0; caller < 2; caller++ {
		go func() {
			defer callers.Done()
			results <- application.Shutdown("TEST", time.Second)
		}()
	}
	callers.Wait()
	close(results)
	for shutdownErr := range results {
		if shutdownErr != nil {
			t.Fatalf("shutdown failed: %v", shutdownErr)
		}
	}
	if got := <-holdDone; got != "held" {
		t.Fatalf("active request was not drained: %s", got)
	}

	metrics := application.Metrics()
	if metrics.Ready || metrics.ActiveRequests != 0 || metrics.CleanupRuns != 1 || metrics.WorkerActive || metrics.Listening {
		t.Fatalf("unexpected final metrics: %+v", metrics)
	}
	if connection, dialErr := net.DialTimeout("tcp", address.String(), 100*time.Millisecond); dialErr == nil {
		_ = connection.Close()
		t.Fatal("listener remained reachable")
	}

	events := application.Events()
	names := make([]string, 0, len(events))
	for _, event := range events {
		names = append(names, event.Event)
	}
	want := []string{"server_ready", "shutdown_started", "resource_cleanup_completed", "shutdown_completed"}
	if !reflect.DeepEqual(names, want) {
		t.Fatalf("unexpected events: got=%v want=%v", names, want)
	}
}

func waitUntil(t *testing.T, predicate func() bool) {
	t.Helper()
	deadline := time.Now().Add(2 * time.Second)
	for !predicate() {
		if time.Now().After(deadline) {
			t.Fatal("condition was not reached")
		}
		time.Sleep(10 * time.Millisecond)
	}
}

func responseStatus(response *http.Response) any {
	if response == nil {
		return nil
	}
	return response.StatusCode
}
